File: marginalia-app/src/types/document.ts
import type { Suggestion } from "./suggestion";

export type DocumentSource = "Local" | "GoogleDocs";

export type DocumentStatus = "Draft" | "Analyzed";

export interface TextRange {
  start: number;
  end: number;
}

export interface Document {
  id: string;
  userId: string;
  filename: string;
  source: DocumentSource;
  content: string;
  title: string;
  status: DocumentStatus;
  createdAt: string;
  updatedAt: string;
  suggestions: Suggestion[];
}

export interface DocumentSummary {
  id: string;
  title: string;
  filename: string;
  source: DocumentSource;
  status: DocumentStatus;
  createdAt: string;
  updatedAt: string;
  suggestionCount: number;
}

File: marginalia-service/src/Domain/Models/Document.cs
using System.Text.Json.Serialization;

namespace Marginalia.Domain.Models;

/// <summary>
/// A manuscript or document uploaded by the author for editorial analysis.
/// </summary>
public sealed record Document
{
    [JsonPropertyName("id")]
    public required string Id { get; init; }

    [JsonPropertyName("userId")]
    public string UserId { get; init; } = "_anonymous";

    [JsonPropertyName("filename")]
    public required string Filename { get; init; }

    [JsonPropertyName("source")]
    public required DocumentSource Source { get; init; }

    [JsonPropertyName("title")]
    public string Title { get; init; } = "";

    [JsonPropertyName("status")]
    public DocumentStatus Status { get; init; } = DocumentStatus.Draft;

    [JsonPropertyName("createdAt")]
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.MinValue;

    [JsonPropertyName("updatedAt")]
    public DateTimeOffset UpdatedAt { get; init; } = DateTimeOffset.MinValue;

    [JsonPropertyName("content")]
    public required string Content { get; init; }

    [JsonPropertyName("suggestions")]
    public IReadOnlyList<Suggestion> Suggestions { get; init; } = [];
}
